### ISSUE TITLE
Spurious error with constrained type variables in generic functions

### ISSUE THREAD

**Issue by whoiscc** [python/mypy#5493]
Reproduction:

```python
from typing import TypeVar, List


class A:
    pass


class B(A):
    pass


class C:
    pass


LooseType = TypeVar('LooseType', A, B, C)
ListType = List[LooseType]


def foo(xs: ListType[LooseType]) -> None:
    pass


StrictType = TypeVar('StrictType', B, C)


def bar(xs: ListType[StrictType]) -> None:
    foo(xs)
```

This is denied with error reporting:

```
test.py:28: error: Argument 1 to "foo" has incompatible type "List[B]"; expected "List[A]"
test.py:28: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
test.py:28: note: Consider using "Sequence" instead, which is covariant
```

Is this really not allowed? If `B` is not derived from `A`, then everything is ok. 

In my original code, the container of `A`, `B` and `C` is `Dict[List[Tuple[LooseType, str]]]`. I cannot change `List` to `Sequence` because I need `append` method.

In addition, the notes disappear in complicated case.

**Comment by ilevkivskyi** [comment]
Thanks for reporting!

Here is a simpler repro (that also is cleaner, since your original repro uses some questionable patterns):
```python
class A:
    pass
class B(A):
    pass
class C:
    pass

LooseType = TypeVar('LooseType', A, B, C)
def foo(xs: List[LooseType]) -> LooseType:
    return xs[0]


StrictType = TypeVar('StrictType', B, C)
def bar(xs: List[StrictType]) -> StrictType:
    foo(xs)
    return xs[0]
```
It seems to me (but this is a pure guess). This is subtle combination of https://github.com/python/mypy/issues/4872 and how functions with restricted type variables are checked.

**Comment by whoiscc** [comment]
Actually I uses `StrictType` and `LooseType` in such way:

```python
from typing import TypeVar, List


class A:
    pass


class B(A):
    pass


class C:
    pass


LooseType = TypeVar('LooseType', A, B, C)
ListType = List[LooseType]


def foo(xs: ListType[LooseType]) -> LooseType:
    return xs[0]


StrictType = TypeVar('StrictType', B, C)


def bar(xs: ListType[StrictType]) -> StrictType:
    return foo(xs)
```

(Only last two lines are changed.)

And this causes one more error with inheritance:

```
test.py:28: error: Incompatible return value type (got "A", expected "B")
```

It does not exist neither if `B` is not derived from `A`. I thought this is an unrelevant issue and eliminated it from original snippet.
